vertical-pod-autoscaler/recommender/cluster/spec_client_test_util.go

```diff
@@ -0,0 +1,154 @@
+	containerSpec12 := newTestContainerSpec(podID1, "Name12", 1000, 1024)
+	containerSpec21 := newTestContainerSpec(podID2, "Name21", 2000, 2048)
+	containerSpec22 := newTestContainerSpec(podID2, "Name22", 4000, 4096)
+
+	testCase.podSpecs = append(testCase.podSpecs, newTestPodSpec(podID1, containerSpec11, containerSpec12))
+	testCase.podSpecs = append(testCase.podSpecs, newTestPodSpec(podID2, containerSpec21, containerSpec22))
+
+	return testCase
+}
+func newTestContainerSpec(podID model.PodID, containerName string, milicores int, memory int) model.BasicContainerSpec {
+	containerID := model.ContainerID{
+		PodID:         podID,
+		ContainerName: containerName,
+	}
+	requestedResources := map[model.MetricName]model.ResourceAmount{
+		model.ResourceCPU:    model.ResourceAmount(milicores),
+		model.ResourceMemory: model.ResourceAmount(memory),
+	}
+	return model.BasicContainerSpec{
+		ID:      containerID,
+		Image:   containerName + "Image",
+		Request: requestedResources,
+	}
+}
+
+func newTestPodSpec(podId model.PodID, containerSpecs ...model.BasicContainerSpec) *model.BasicPodSpec {
+	return &model.BasicPodSpec{
+		ID:         podId,
+		PodLabels:  map[string]string{podId.PodName + "LabelKey": podId.PodName + "LabelValue"},
+		Containers: containerSpecs,
+	}
+}
+
+func (tc *specClientTestCase) createFakeSpecClient() SpecClient {
+	podListerMock := new(podListerMock)
+	podListerMock.On("List").Return(tc.getFakePods(), nil)
+
+	return NewSpecClient(podListerMock)
+}
+
+func (tc *specClientTestCase) getFakePods() []*v1.Pod {
+	pods := []*v1.Pod{}
+	for _, podSpec := range tc.podSpecs {
+		pods = append(pods, newPod(podSpec))
+	}
+	return pods
+}
+
+func newPod(podSpec *model.BasicPodSpec) *v1.Pod {
+
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:         podSpec.ID.Namespace,
+			Name:              podSpec.ID.PodName,
+			CreationTimestamp: metav1.Time{time.Now()},
+			Labels:            podSpec.PodLabels,
+		},
+		Spec: v1.PodSpec{
+			Containers: make([]v1.Container, len(podSpec.Containers)),
+		},
+	}
+
+	for i, containerSpec := range podSpec.Containers {
+		pod.Spec.Containers[i] = v1.Container{
+			Name:  containerSpec.ID.ContainerName,
+			Image: containerSpec.Image,
+			Resources: v1.ResourceRequirements{
+				Requests: calculateResourceList(containerSpec.Request),
+			},
+		}
+	}
+	return pod
+}
+
+func calculateResourceList(usage map[model.MetricName]model.ResourceAmount) k8sapiv1.ResourceList {
+	cpuCores := big.NewRat(int64(usage[model.ResourceCPU]), 1000)
+	cpuQuantityString := cpuCores.FloatString(3)
+
+	memoryBytes := big.NewInt(int64(usage[model.ResourceMemory]))
+	memoryQuantityString := memoryBytes.String()
+
+	resourceMap := map[k8sapiv1.ResourceName]resource.Quantity{
+		k8sapiv1.ResourceCPU:    resource.MustParse(cpuQuantityString),
+		k8sapiv1.ResourceMemory: resource.MustParse(memoryQuantityString),
+	}
+	return k8sapiv1.ResourceList(resourceMap)
```

Comment:
This looks a bit convoluted: you're implementing a whole logic for translating the model objects back to the API objects (which is actually more complex than the original code that you're testing).

Wouldn't it be more natural and readable to e.g. parse API objects from a yaml string (see e.g. https://stackoverflow.com/questions/44306554/how-to-deserialize-kubernetes-yaml-file). Then I think you also have more chance of getting correctly constructed API objects. WDYT?

Reply:
Seems like an interesting idea, thanks for the suggestion.
I looked at this shortly and was able to parse yaml, but had some issues with converting obj to `v1.Pod` type. Will look into that a little bit more, and if successful, include in the next commit

Reply:
implemented. Should be simpler now. Thanks again for the idea!